# Daniel Michell Pérez Ruiz - Machine Learning I


# 2. Serie Pandas
Está es la segunda de tres series de módulos científicos.

Pandas es un módulo que se usa para trabajar los datos como dataFrames, el próposito de esta serie es que se familiaricen con el objeto tipo DataFrame y practiquen la exportación de datos.

Saludos.

In [1]:
import pandas as pd
import numpy as np


1) Considerando el siguiente set de datos:

       a) Crea un dataframe a partir de datos
       b) Despliega los primeros tres datos de dicho dataframe
       c) Despliga los últimos dos datos de dicho dataframe
       d) Una descripción de los datos (count, mean, std, ...) 

In [2]:
datos = {
    'fecha':pd.date_range('20200430', periods=6),
    'calidad':pd.Categorical(['buena', 'muy buena', 'mala', 'buena', 'muy mala', 'buena'],
                                 categories = ['muy mala', 'mala', 'regular', 'buena', 'muy buena'],
                                 ordered = True),
    'asistentes':[10, 25, 12, 3, 23, 43],
    'recomendaciones':['hablar más lento', 'me aburri', 'Derek hacia cosas raras', 'excelente', 'no se entendio nada',
               '¿Y el recreo?']
}

#Creamos el dataframe a partir de escribir pd.DataFrame(objeto a crear)
df = pd.DataFrame(datos)

print(df)

       fecha    calidad  asistentes          recomendaciones
0 2020-04-30      buena          10         hablar más lento
1 2020-05-01  muy buena          25                me aburri
2 2020-05-02       mala          12  Derek hacia cosas raras
3 2020-05-03      buena           3                excelente
4 2020-05-04   muy mala          23      no se entendio nada
5 2020-05-05      buena          43            ¿Y el recreo?


In [3]:
#Desplegamos los primeros datos con la palabra head
df.head(3)

fecha    calidad  asistentes          recomendaciones
0 2020-04-30      buena          10         hablar más lento
1 2020-05-01  muy buena          25                me aburri
2 2020-05-02       mala          12  Derek hacia cosas raras

In [4]:
#Desplegamos los últimos datos con la palabra tail
df.tail(2)

fecha   calidad  asistentes      recomendaciones
4 2020-05-04  muy mala          23  no se entendio nada
5 2020-05-05     buena          43        ¿Y el recreo?

In [5]:
#Describimos los datos con describe
df.describe()

asistentes
count    6.000000
mean    19.333333
std     14.236104
min      3.000000
25%     10.500000
50%     17.500000
75%     24.500000
max     43.000000

2) Crea una función que pida número de columnas, su nombre y número de datos a ingresar de manera que el usuario pueda crear un dataframe con los datos de su preferencia.

In [6]:
########## DATAFRAME SOBRE DEPORTISTAS ###########################


def dataframe(NumCol,NumDatos):
    Names = []
    Datos = []
    df = pd.DataFrame()
    Matriz = np.empty((NumDatos,NumCol),dtype=object)
    #Lista = []
    columna = 0
    renglon = 0
    for x in range(0,NumCol):
        columna +=1
        Name = input('Escribe el nombre de la columna: ' )
        Names.append(Name)
        print('Columna ',columna )
    for y in range(0,NumDatos):
        #renglon +=1
        #Dato = input('Escribe el nombre de los renglones:')
        #Datos.append(Dato)
        #print('Renglon', renglon)
        
        renglon +=1
        #Dato = input('Escribe el nombre de los renglones:')
        Datos.append(renglon)
        #print('Renglon', renglon)
    
    for r in range(len(Datos)):
        for c in range(len(Names)):
            Matriz[r][c] = (input('Elemento (%s , %s) de tu DataFrame: ' %(r+1,c+1)))
    
    #Lista = list(Matriz)
    
    
    #Lo podemos dejar así o pasarlo a lista 
    df = pd.DataFrame(Matriz, index = Datos, columns = Names)
    #return Names, Datos, Matriz
        
    return df

        

NumCol = int(input('Cuál es el número de columnas: '))
NumDatos = int(input('Cuál es el número de datos por columna: '))

dataframe(NumCol,NumDatos)


Cuál es el número de columnas: 4
Cuál es el número de datos por columna: 5
Escribe el nombre de la columna: Nombre
Columna  1
Escribe el nombre de la columna: Edad
Columna  2
Escribe el nombre de la columna: Deporte
Columna  3
Escribe el nombre de la columna: Activo/Retirado
Columna  4
Elemento (1 , 1) de tu DataFrame: Serena Williams
Elemento (1 , 2) de tu DataFrame: 38
Elemento (1 , 3) de tu DataFrame: Tennis
Elemento (1 , 4) de tu DataFrame: Activo
Elemento (2 , 1) de tu DataFrame: Lionel Messi
Elemento (2 , 2) de tu DataFrame: 33
Elemento (2 , 3) de tu DataFrame: Futbol
Elemento (2 , 4) de tu DataFrame: Activo
Elemento (3 , 1) de tu DataFrame: Lance Armstrong
Elemento (3 , 2) de tu DataFrame: 48
Elemento (3 , 3) de tu DataFrame: Ciclismo
Elemento (3 , 4) de tu DataFrame: Retirado
Elemento (4 , 1) de tu DataFrame: Tiger Woods
Elemento (4 , 2) de tu DataFrame: 44
Elemento (4 , 3) de tu DataFrame: Golf
Elemento (4 , 4) de tu DataFrame: Retirado
Elemento (5 , 1) de tu DataFrame: Michae

Nombre Edad   Deporte Activo/Retirado
1  Serena Williams   38    Tennis          Activo
2     Lionel Messi   33    Futbol          Activo
3  Lance Armstrong   48  Ciclismo        Retirado
4      Tiger Woods   44      Golf        Retirado
5   Michael Phelps   35  Natación          Activo

In [7]:
################################### DATAFRAME CON ÍNDICES SOBRE CIENTÍFIC@S FAMOS@S ###############
def dataframe(NumCol,NumDatos):
    Names = []
    Datos = []
    df = pd.DataFrame()
    Matriz = np.empty((NumDatos,NumCol),dtype=object)
    #Lista = []
    columna = 0
    renglon = 0
    for x in range(0,NumCol):
        columna +=1
        Name = input('Escribe el nombre de la columna: ' )
        Names.append(Name)
        print('Columna ',columna )
    for y in range(0,NumDatos):
        renglon +=1
        Dato = input('Escribe el nombre de los renglones:')
        Datos.append(Dato)
        print('Renglon', renglon)
        
        #renglon +=1
        #Dato = input('Escribe el nombre de los renglones:')
        #Datos.append(renglon)
        #print('Renglon', renglon)
    
    for r in range(len(Datos)):
        for c in range(len(Names)):
            Matriz[r][c] = (input('Elemento (%s , %s) de tu DataFrame: ' %(r+1,c+1)))
    
    #Lista = list(Matriz)
    
    
    #Lo podemos dejar así o pasarlo a lista 
    df = pd.DataFrame(Matriz, index = Datos, columns = Names)
    #return Names, Datos, Matriz
        
    return df

        

NumCol = int(input('Cuál es el número de columnas: '))
NumDatos = int(input('Cuál es el número de datos por columna: '))

dataframe(NumCol,NumDatos)

Cuál es el número de columnas: 4
Cuál es el número de datos por columna: 5
Escribe el nombre de la columna: Campo de Investigación
Columna  1
Escribe el nombre de la columna: Nacionalidad
Columna  2
Escribe el nombre de la columna: Aportación
Columna  3
Escribe el nombre de la columna: Vida-Muerte
Columna  4
Escribe el nombre de los renglones:Albert Einstein
Renglon 1
Escribe el nombre de los renglones:Isaac Newton
Renglon 2
Escribe el nombre de los renglones:Marie Curie
Renglon 3
Escribe el nombre de los renglones:Charles Darwin
Renglon 4
Escribe el nombre de los renglones:Rosalind Franklin
Renglon 5
Elemento (1 , 1) de tu DataFrame: Física
Elemento (1 , 2) de tu DataFrame: Alemana
Elemento (1 , 3) de tu DataFrame: Teoría de la Relatividad
Elemento (1 , 4) de tu DataFrame: 1879-1955
Elemento (2 , 1) de tu DataFrame: Física
Elemento (2 , 2) de tu DataFrame: Inglesa
Elemento (2 , 3) de tu DataFrame: Leyes de Newton
Elemento (2 , 4) de tu DataFrame: 1643-1727
Elemento (3 , 1) de tu DataF

Campo de Investigación Nacionalidad  \
Albert Einstein                   Física      Alemana   
Isaac Newton                      Física      Inglesa   
Marie Curie               Física/Química       Polaca   
Charles Darwin                  Biología      Inglesa   
Rosalind Franklin                Química      Inglesa   

                                 Aportación Vida-Muerte  
Albert Einstein    Teoría de la Relatividad   1879-1955  
Isaac Newton                Leyes de Newton   1643-1727  
Marie Curie                  Radioactividad   1867-1934  
Charles Darwin            Selección Natural   1809-1882  
Rosalind Franklin  Estructura Molecular ADN   1920-1958


3) A partir del siguiente DataSet realiza:

In [8]:
#Liga del github: https://raw.githubusercontent.com/adrishg/PythonML-PrebecariosProteco/master/DataSets/startupsUSA.csv
#Podrán agregar archivos de github y leerlos usando el url raw
data = pd.read_csv('https://raw.githubusercontent.com/adrishg/Intersemestral_MachineLearning_Proteco2020/master/DataSets/startupsUSA.csv')
data.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

a) Crea una función que reciba un lugar y despligue en el número de empresas que residen en dicho lugar

In [9]:
#Nos dice el número de empresas que se encuentran en cada lugar

def ciudad():
    x = pd.Series(data['State'].value_counts())
    y = input('Dime la ciudad: (California/Florida/New York): ')
    print('La ciudad ', y, 'tiene', x.loc[y]  ,'empresas.' )

ciudad()



Dime la ciudad: (California/Florida/New York): Florida
La ciudad  Florida tiene 16 empresas.


b) Un data set cuyo orden este dado alfabeticamente por los lugares donde están las empresas

In [10]:
#Se ordena alfabeticamente las empresas con sort
by_state = data.sort_values('State')
#by_state.head()
by_state


#Si lo queremos hacer por más de una columna se hace lo siguiente
#by_R_D = data.sort_values(['State','R&D Spend'])


R&D Spend  Administration  Marketing Spend       State     Profit
49       0.00       116983.80         45173.06  California   14681.40
42   23640.93        96189.63        148001.11  California   71498.49
40   28754.33       118546.05        172795.67  California   78239.91
34   46426.07       157693.92        210797.67  California   96712.80
16   78013.11       121597.55        264346.06  California  126992.93
44   22177.74       154806.14         28334.72  California   65200.33
25   64664.71       139553.16        137962.62  California  107404.34
13   91992.39       135495.07        252664.93  California  134307.35
11  100671.96        91790.61        249744.55  California  144259.40
20   76253.86       113867.30        298664.47  California  118474.03
39   38558.51        82982.09        174999.30  California   81005.76
6   134615.46       147198.87        127716.82  California  156122.51
37   44069.95        51283.14        197029.42  California   89949.14
32   63408.86       129219.61         46085.25  California   97427.84
47       0.00       135426.92             0.00  California   42559.73
1   162597.70       151377.59        443898.53  California  191792.06
9   123334.88       108679.17        304981.62  California  149759.96
28   66051.52       182645.56        118148.20     Florida  103282.38
26   75328.87       144135.98        134050.07     Florida  105733.54
30   61994.48       115641.28         91131.24     Florida   99937.59
33   55493.95       103057.49        214634.81     Florida   96778.92
23   67532.53       105751.03        304768.73     Florida  108733.99
36   28663.76       127056.21        201126.82     Florida   90708.19
41   27892.92        84710.77        164470.71     Florida   77798.83
18   91749.16       114175.79        294919.57     Florida  124266.90
14  119943.24       156547.42        256512.92     Florida  132602.65
12   93863.75       127320.38        249839.44     Florida  141585.52
10  101913.08       110594.11        229160.95     Florida  146121.95
7   130298.13       145530.06        323876.68     Florida  155752.60
46    1315.46       115816.21        297114.46     Florida   49490.75
4   142107.34        91391.77        366168.42     Florida  166187.94
2   153441.51       101145.55        407934.54     Florida  191050.39
22   73994.56       122782.75        303319.26     Florida  110352.25
45    1000.23       124153.04          1903.93    New York   64926.08
38   20229.59        65947.93        185265.10    New York   81229.06
43   15505.73       127382.30         35534.17    New York   69758.98
0   165349.20       136897.80        471784.10    New York  192261.83
31   61136.38       152701.92         88218.23    New York   97483.56
29   65605.48       153032.06        107138.38    New York  101004.64
27   72107.60       127864.55        353183.81    New York  105008.31
48     542.05        51743.15             0.00    New York   35673.41
21   78389.47       153773.43        299737.29    New York  111313.02
19   86419.70       153514.11             0.00    New York  122776.86
17   94657.16       145077.58        282574.31    New York  125370.37
15  114523.61       122616.84        261776.23    New York  129917.04
8   120542.52       148718.95        311613.29    New York  152211.77
5   131876.90        99814.71        362861.36    New York  156991.12
3   144372.41       118671.85        383199.62    New York  182901.99
35   46014.02        85047.44        205517.64    New York   96479.51
24   77044.01        99281.34        140574.81    New York  108552.04

c) Una función donde ingreses el index númerico y despliegue dichos datos de manera legible:

    Ejemplo: [In]: 0  [Out]: La empresa que reside en New York tiene un gasto de Marketing de 471784.10 dólares mientras que sus ganancias son 192261.83 dólares...

In [11]:
#Nos despliega la información de la empresa requerida 
def index(i):
    print('La empresa que reside en ', data.iloc[i,3], 
          'tiene un gasto en Marketing de ', data.iloc[i,2], 'dolares mientras que sus ganancias son ', 
          data.iloc[i,4], 'dolares.')

    
index(2)

La empresa que reside en  Florida tiene un gasto en Marketing de  407934.54 dolares mientras que sus ganancias son  191050.39 dolares.



4) Deberán cargar los datos de iris setosa de dos maneras:

a) Deberán dirigirse al repositorio del curso: https://github.com/adrishg/Intersemestral_MachineLearning_Proteco2020
De ahí a la carpeta de DataSets, ahí encontrarán el archivo iris.csv deben darle click para previsualizar el archivo. Ahí encontrarán un botón que dice Raw denle click y copien la liga, ese es el link raw y podrán exportarlo como lo hice en la parte de arriba con los datos de las empresas estadounidenses.
Desplieguen los datos con un describe para que revisen que se exporto bien.


In [12]:
data = pd.read_csv('https://raw.githubusercontent.com/adrishg/Intersemestral_MachineLearning_Proteco2020/master/DataSets/iris.csv')
data.head()

Id  sepal_length  sepal_width  petal_length  petal_width      species
0   1           5.1          3.5           1.4          0.2  Iris-setosa
1   2           4.9          3.0           1.4          0.2  Iris-setosa
2   3           4.7          3.2           1.3          0.2  Iris-setosa
3   4           4.6          3.1           1.5          0.2  Iris-setosa
4   5           5.0          3.6           1.4          0.2  Iris-setosa

In [13]:
#Revisamos que se hallan exportado correctamente
data.describe()

Id  sepal_length  sepal_width  petal_length  petal_width
count  150.000000    150.000000   150.000000    150.000000   150.000000
mean    75.500000      5.843333     3.054000      3.758667     1.198667
std     43.445368      0.828066     0.433594      1.764420     0.763161
min      1.000000      4.300000     2.000000      1.000000     0.100000
25%     38.250000      5.100000     2.800000      1.600000     0.300000
50%     75.500000      5.800000     3.000000      4.350000     1.300000
75%    112.750000      6.400000     3.300000      5.100000     1.800000
max    150.000000      7.900000     4.400000      6.900000     2.500000

b) El dataset de Iris es muy conocido en el área ya que está precargado en sklearn.datasets. Hay varias formas de hacerlo, una de ellas se las puse abajo, deben comentar el código para que sepan que hace.

In [14]:
from sklearn.datasets import load_iris                         #De la paquetería de sklearn la cual contiene muchísimos datos, 
                                                              #importamos un set de datos llamado iris (sklearn es una de las paquetrías más usadas para ML)
data = load_iris()                                            #Al set obtenido le asignamos una variable
irisDataFrame = pd.DataFrame(data.data, columns=data.feature_names)  #A la variable "data" lo convertimos en un DF y le asignamos el nombre de sus columnas
irisDataFrame.head()                                        #Desplegamos el encabezado de dicho DF

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

c) Busquen otros datasets que se encuentren en sklearn y carguen uno.

In [15]:
# un data set el cual evalua 11 tipos variables para obtener la calidad del vino
from sklearn.datasets import load_wine
data = load_wine()
vino = pd.DataFrame(data.data, columns=data.feature_names)
vino.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0


5) En esta opción eres libre de poner cualquier función que use elementos y funciones de la librería pandas. El requerimiento es que uses iloc, puedes usar alguno de los datasets cargados en esta libreta de jupyter.

In [16]:
#Dataset para concatenar dos dataset


# Se crea un nuevo dataset
States = ['NY', 'NY', 'NY', 'NY', 'FL', 'FL', 'GA', 'GA', 'FL', 'FL'] #Le asignamos los estados 
data = [1.0, 2, 3, 4, 5, 6, 7, 8, 9, 10]                              #Los datos del df
idx = pd.date_range('1/1/2012', periods=10, freq='MS')                #Los índices ordenados por fechas
df1 = pd.DataFrame(data, index=idx, columns=['Revenue'])              #Creamos el df
df1['State'] = States                                                 #Agregamos otra columna llamada 'State' al df

#Se crea el segundo dataframe con similares instrucciones
data2 = [10.0, 10.0, 9, 9, 8, 8, 7, 7, 6, 6]
idx2 = pd.date_range('1/1/2013', periods=10, freq='MS')
df2 = pd.DataFrame(data2, index=idx2, columns=['Revenue'])
df2['State'] = States

In [17]:
#Imprimimos por separado para observar la información 
df1

Revenue State
2012-01-01      1.0    NY
2012-02-01      2.0    NY
2012-03-01      3.0    NY
2012-04-01      4.0    NY
2012-05-01      5.0    FL
2012-06-01      6.0    FL
2012-07-01      7.0    GA
2012-08-01      8.0    GA
2012-09-01      9.0    FL
2012-10-01     10.0    FL

In [18]:
#Imprimimos por separado
df2

Revenue State
2013-01-01     10.0    NY
2013-02-01     10.0    NY
2013-03-01      9.0    NY
2013-04-01      9.0    NY
2013-05-01      8.0    FL
2013-06-01      8.0    FL
2013-07-01      7.0    GA
2013-08-01      7.0    GA
2013-09-01      6.0    FL
2013-10-01      6.0    FL

In [19]:
#Concatemamos los df
df = pd.concat([df1,df2])
df

Revenue State
2012-01-01      1.0    NY
2012-02-01      2.0    NY
2012-03-01      3.0    NY
2012-04-01      4.0    NY
2012-05-01      5.0    FL
2012-06-01      6.0    FL
2012-07-01      7.0    GA
2012-08-01      8.0    GA
2012-09-01      9.0    FL
2012-10-01     10.0    FL
2013-01-01     10.0    NY
2013-02-01     10.0    NY
2013-03-01      9.0    NY
2013-04-01      9.0    NY
2013-05-01      8.0    FL
2013-06-01      8.0    FL
2013-07-01      7.0    GA
2013-08-01      7.0    GA
2013-09-01      6.0    FL
2013-10-01      6.0    FL